File: lib/services/historico_service.dart
import 'package:fluire/models/registro_auditoria.dart';
import 'package:fluire/services/alunos_service.dart';
import 'package:fluire/services/aulas_service.dart';
import 'package:fluire/services/usuarios_service.dart';

class HistoricoService {
  final AlunosService _alunosService = AlunosService();
  final AulasService _aulasService = AulasService();
  final UsuariosService _usuariosService = UsuariosService();

  Future<List<RegistroAuditoria>> listarAtividades({int limite = 50}) async {
    // Busca alunos, aulas e usuários em paralelo
    final resultados = await Future.wait([
      _alunosService.listar(),
      _aulasService.listar(),
      _carregarUsuarios(),
    ]);

    final alunos = resultados[0] as List;
    final aulas = resultados[1] as List;
    final mapaUsuarios = resultados[2] as Map<int, String>;

    final registros = <RegistroAuditoria>[];

    for (final aluno in alunos) {
      if (aluno.createdAt != null || aluno.createdBy != null) {
        registros.add(
          RegistroAuditoria(
            entidade: TipoEntidadeAuditoria.aluno,
            acao: TipoAcaoAuditoria.criacao,
            titulo: aluno.nome,
            subtitulo: aluno.email.isNotEmpty ? aluno.email : aluno.telefone,
            criadoPor: _nomeUsuario(aluno.createdBy, mapaUsuarios),
            dataCriacao: aluno.createdAt,
          ),
        );
      }

      if (aluno.updatedAt != null || aluno.updatedBy != null) {
        registros.add(
          RegistroAuditoria(
            entidade: TipoEntidadeAuditoria.aluno,
            acao: TipoAcaoAuditoria.atualizacao,
            titulo: aluno.nome,
            subtitulo: aluno.email.isNotEmpty ? aluno.email : aluno.telefone,
            criadoPor: _nomeUsuario(aluno.createdBy, mapaUsuarios),
            atualizadoPor: _nomeUsuario(aluno.updatedBy, mapaUsuarios),
            dataCriacao: aluno.createdAt,
            dataAtualizacao: aluno.updatedAt,
          ),
        );
      }
    }

    for (final aula in aulas) {
      if (aula.createdAt != null || aula.createdBy != null || aula.id.isNotEmpty) {
        registros.add(
          RegistroAuditoria(
            entidade: TipoEntidadeAuditoria.aula,
            acao: TipoAcaoAuditoria.criacao,
            titulo: aula.nome,
            subtitulo: aula.horario,
            criadoPor: _nomeUsuario(aula.createdBy, mapaUsuarios) ?? 'Não informado',
            dataCriacao: aula.createdAt ?? DateTime.now(),
          ),
        );
      }

      // Se temos indicação de alteração ou se pudermos identificar que houve alteração (updatedBy/updatedAt)
      if (aula.updatedAt != null || aula.updatedBy != null) {
        registros.add(
          RegistroAuditoria(
            entidade: TipoEntidadeAuditoria.aula,
            acao: TipoAcaoAuditoria.atualizacao,
            titulo: aula.nome,
            subtitulo: aula.horario,
            criadoPor: _nomeUsuario(aula.createdBy, mapaUsuarios),
            atualizadoPor: _nomeUsuario(aula.updatedBy, mapaUsuarios) ?? 'Usuário logado',
            dataCriacao: aula.createdAt,
            dataAtualizacao: aula.updatedAt ?? DateTime.now(),
          ),
        );
      }
    }

    registros.sort((a, b) {
      final da = a.dataReferencia ?? DateTime.fromMillisecondsSinceEpoch(0);
      final db = b.dataReferencia ?? DateTime.fromMillisecondsSinceEpoch(0);
      return db.compareTo(da);
    });

    if (registros.length > limite) {
      return registros.sublist(0, limite);
    }
    return registros;
  }

  /// Carrega a lista de usuários e retorna um mapa ID → Nome.
  /// Em caso de falha, retorna mapa vazio para não quebrar a tela.
  Future<Map<int, String>> _carregarUsuarios() async {
    try {
      final usuarios = await _usuariosService.listar();
      final mapa = <int, String>{};
      for (final usuario in usuarios) {
        final idInt = int.tryParse(usuario.id);
        if (idInt != null && usuario.nome.isNotEmpty) {
          mapa[idInt] = usuario.nome;
        }
      }
      return mapa;
    } catch (_) {
      return <int, String>{};
    }
  }

  /// Resolve o ID de usuário para o nome real.
  /// Retorna null se o ID for nulo (campo não preenchido).
  String? _nomeUsuario(dynamic id, Map<int, String> mapa) {
    if (id == null) return null;
    final idInt = id is int ? id : int.tryParse(id.toString());
    if (idInt == null) return null;
    return mapa[idInt] ?? 'Usuário desconhecido';
  }
}

File: src/services/historicoService.js
import * as alunosService from './alunosService';
import * as aulasService from './aulasService';
import * as usuariosService from './usuariosService';

export async function listarAtividades(limite = 50) {
  const [alunos, aulas, mapaUsuarios] = await Promise.all([
    alunosService.listar(),
    aulasService.listar(),
    carregarUsuarios(),
  ]);

  const registros = [];

  for (const aluno of alunos) {
    if (aluno.created_at || aluno.created_by) {
      registros.push({
        entidade: 'aluno',
        acao: 'criacao',
        titulo: aluno.nome,
        subtitulo: aluno.email || aluno.telefone || '',
        criadoPor: nomeUsuario(aluno.created_by, mapaUsuarios),
        dataCriacao: parseData(aluno.created_at),
      });
    }
    if (aluno.updated_at || aluno.updated_by) {
      registros.push({
        entidade: 'aluno',
        acao: 'atualizacao',
        titulo: aluno.nome,
        subtitulo: aluno.email || aluno.telefone || '',
        criadoPor: nomeUsuario(aluno.created_by, mapaUsuarios),
        atualizadoPor: nomeUsuario(aluno.updated_by, mapaUsuarios),
        dataCriacao: parseData(aluno.created_at),
        dataAtualizacao: parseData(aluno.updated_at),
      });
    }
  }

  for (const aula of aulas) {
    if (aula.createdAt || aula.createdBy || aula.id) {
      registros.push({
        entidade: 'aula',
        acao: 'criacao',
        titulo: aula.nome,
        subtitulo: `${aula.horarioInicio} - ${aula.horarioFim}`,
        criadoPor: nomeUsuario(aula.createdBy, mapaUsuarios) || 'Não informado',
        dataCriacao: parseData(aula.createdAt) || new Date(),
      });
    }
    if (aula.updatedAt || aula.updatedBy) {
      registros.push({
        entidade: 'aula',
        acao: 'atualizacao',
        titulo: aula.nome,
        subtitulo: `${aula.horarioInicio} - ${aula.horarioFim}`,
        criadoPor: nomeUsuario(aula.createdBy, mapaUsuarios),
        atualizadoPor: nomeUsuario(aula.updatedBy, mapaUsuarios) || 'Usuário logado',
        dataCriacao: parseData(aula.createdAt),
        dataAtualizacao: parseData(aula.updatedAt) || new Date(),
      });
    }
  }

  registros.sort((a, b) => {
    const da = dataReferencia(a) || new Date(0);
    const db = dataReferencia(b) || new Date(0);
    return db - da;
  });

  return registros.slice(0, limite);
}

function dataReferencia(registro) {
  return registro.acao === 'criacao' ? registro.dataCriacao : registro.dataAtualizacao;
}

async function carregarUsuarios() {
  try {
    const usuarios = await usuariosService.listar();
    const mapa = {};
    for (const u of usuarios) {
      const id = parseInt(u.id, 10);
      if (!isNaN(id) && u.nome) mapa[id] = u.nome;
    }
    return mapa;
  } catch {
    return {};
  }
}

function nomeUsuario(id, mapa) {
  if (id == null) return null;
  const idInt = parseInt(id, 10);
  if (isNaN(idInt)) return null;
  return mapa[idInt] || 'Usuário desconhecido';
}

function parseData(value) {
  if (!value) return null;
  if (value instanceof Date) return value;
  const d = new Date(value);
  return isNaN(d.getTime()) ? null : d;
}

export function formatarData(data) {
  if (!data) return '—';
  const d = data instanceof Date ? data : new Date(data);
  if (isNaN(d.getTime())) return '—';
  const dia = String(d.getDate()).padStart(2, '0');
  const mes = String(d.getMonth() + 1).padStart(2, '0');
  const ano = d.getFullYear();
  const hora = String(d.getHours()).padStart(2, '0');
  const min = String(d.getMinutes()).padStart(2, '0');
  return `${dia}/${mes}/${ano} às ${hora}:${min}`;
}
